lib/bridge/PowerlevelSyncer.js

```diff
@@ -0,0 +1,402 @@
+const log = require("../logging").get("PowerlevelSyncer");
+
+const MODES_TO_WATCH = [
+    "m", // This channel is "moderated" - only voiced users can speak.
+         // We want to ensure we do not miss rooms that get unmoderated.
+    "k", // keylock - needs a password
+    "i", // invite only
+    "s", // secret - don't show in channel lisitings
+];
+
+const PRIVATE_MODES = [
+    "k",
+    "i",
+    "s",
+]
+
+/**
+ * This class is supplimentary to the IrcHandler class. This
+ * class handles incoming mode changes as well as computing the new
+ * power level state.
+ */
+class PowerlevelSyncer {
+
+    constructor(ircBridge) {
+        this._ircBridge = ircBridge;
+        // Warning: This cache is currently unbounded.
+        this._powerLevelsForRoom = {
+            // roomId:PowerLevelObject
+        };
+    }
+
+    /**
+     * Called when a m.room.power_levels is forwarded to the bridge. This should
+     * happen when a Matrix user or the bridge changes the power levels for a room.
+     * @param {MatrixEvent} event The matrix event.
+     */
+    onMatrixPowerlevelEvent(event) {
+        this._powerLevelsForRoom[event.room_id] = event.content;
```

Comment:
I believe this is a js-sdk Event class, so we should be using the functions to pull the content out.
```suggestion
        this._powerLevelsForRoom[event.room_id] = event.getContent();
```

Reply:
It's not, and it's terribly misleading. This event comes from an appservice transaction which we've unwrapped in `IrcBridge`.

Reply:
ohgod.gif